lib/experimental/modules/class-gutenberg-modules.php

```diff
@@ -0,0 +1,195 @@
+	 */
+	public static function register( $module_identifier, $src, $dependencies = array(), $version = false ) {
+		// Register the module if it's not already registered.
+		if ( ! isset( self::$registered[ $module_identifier ] ) ) {
+			$deps = array(
+				'static'  => isset( $dependencies['static'] ) || isset( $dependencies['dynamic'] ) ? $dependencies['static'] ?? array() : $dependencies,
+				'dynamic' => isset( $dependencies['dynamic'] ) ? $dependencies['dynamic'] : array(),
+			);
+
+			self::$registered[ $module_identifier ] = array(
+				'src'          => $src,
+				'version'      => $version,
+				'dependencies' => $deps,
+			);
+		}
+	}
+
+	/**
+	 * Enqueues a module in the page.
+	 *
+	 * @param string $module_identifier The identifier of the module.
+	 */
+	public static function enqueue( $module_identifier ) {
+		// Add the module to the queue if it's not already there.
+		if ( ! in_array( $module_identifier, self::$enqueued, true ) ) {
+			self::$enqueued[] = $module_identifier;
+		}
+	}
+
+	/**
+	 * Returns the import map array.
+	 *
+	 * @return array Associative array with 'imports' key mapping to an array of module identifiers and their respective source strings.
+	 */
+	public static function get_import_map() {
+		$imports = array();
+		foreach ( self::get_dependencies( self::$enqueued, array( 'static', 'dynamic' ) ) as $module_identifier => $module ) {
+			$imports[ $module_identifier ] = $module['src'] . self::get_version_query_string( $module['version'] );
+		}
+		return array( 'imports' => $imports );
+	}
+
+	/**
+	 * Prints the import map.
+	 */
+	public static function print_import_map() {
+		$import_map = self::get_import_map();
+		if ( ! empty( $import_map['imports'] ) ) {
+			echo '<script type="importmap">' . wp_json_encode( self::get_import_map(), JSON_HEX_TAG | JSON_HEX_AMP ) . '</script>';
+		}
+	}
+
+	/**
+	 * Prints all the enqueued modules using <script type="module">.
+	 */
+	public static function print_enqueued_modules() {
+		foreach ( self::$enqueued as $module_identifier ) {
+			if ( isset( self::$registered[ $module_identifier ] ) ) {
+				$module = self::$registered[ $module_identifier ];
+				wp_print_script_tag(
+					array(
+						'type' => 'module',
+						'src'  => $module['src'] . self::get_version_query_string( $module['version'] ),
+						'id'   => $module_identifier,
+					)
+				);
+			}
+		}
+	}
+
+	/**
+	 * Prints the link tag with rel="modulepreload" for all the static
+	 * dependencies of the enqueued modules.
+	 */
+	public static function print_module_preloads() {
+		foreach ( self::get_dependencies( self::$enqueued, array( 'static' ) ) as $dependency_identifier => $module ) {
+				echo '<link rel="modulepreload" href="' . $module['src'] . self::get_version_query_string( $module['version'] ) . '" id="' . $dependency_identifier . '">';
```

Comment:
This id would be duplicated with the `print_enqueued_modules`. It would end up printing:

`<link rel="modulepreload" href="module_1_src?ver=1.0.0" id="module_1">`
`<script type="module" src="module_1_src?ver=1.0.0" id="module_1">`

is that ok?

Reply:
Good catch. The preload is unnecessary in that case, so we have to filter the enqueued modules out of the preloaded ones.

Reply:
I've added it to the follow-up tasks in the PR description. I'll add it to the Tracking Issue as well once/if this is merged.